src/datasets/utils/py_utils.py

```diff
@@ -318,19 +318,52 @@ def _single_map_nested(args):
 
 
 def map_nested(
-    function,
-    data_struct,
+    function: Callable[[Any], Any],
+    data_struct: Any,
     dict_only: bool = False,
     map_list: bool = True,
     map_tuple: bool = False,
     map_numpy: bool = False,
     num_proc: Optional[int] = None,
-    types=None,
+    multiprocessing_min_length: int = 2,
+    types: Optional[tuple] = None,
     disable_tqdm: bool = True,
     desc: Optional[str] = None,
-):
+) -> Any:
     """Apply a function recursively to each element of a nested data struct.
-    If num_proc > 1 and the length of data_struct is longer than num_proc: use multi-processing
+
+    Use multiprocessing if num_proc > 1 and the length of data_struct is greater than or equal to
+    `multiprocessing_min_length`.
+
+    <Changed version="2.4.0">
+
+    Before version 2.4.0, multiprocessing was not used if `num_proc` was greater than or equal to ``len(iterable)``.
+
+    Now, if `num_proc` is greater than or equal to ``len(iterable)``, `num_proc` is set to ``len(iterable)`` and
+    multiprocessing is used.
+
+    </Changed>
+
+    Args:
+        function (`Callable`): Function to be applied to `data_struct`.
+        data_struct (`Any`): Data structure to apply `function` to.
+        dict_only (`bool`, default `False`): Whether only apply `function` recursively to `dict` values in
+            `data_struct`.
+        map_list (`bool`, default `True`): Whether also apply `function` recursively to `list` elements (besides `dict`
+            values).
+        map_tuple (`bool`, default `False`): Whether also apply `function` recursively to `tuple` elements (besides
+            `dict` values).
+        map_numpy (`bool, default `False`): Whether also apply `function` recursively to `numpy.array` elements (besides
+            `dict` values).
+        num_proc (`int`, *optional*): Number of processes.
+        multiprocessing_min_length (`int`, default `2`): Minimum length of `data_struct` required for multiprocessing.
```

Comment:
Maybe name it `min_length_for_parallelism` in case we ever want to add support for multithreading in map_nested ?

Reply:
@lhoestq I renamed it to `parallel_min_length` so that its naming aligns with other variable names:
- `IN_MEMORY_MAX_SIZE`
- `STREAMING_READ_MAX_RETRIES`

I just replaced `parallelism` with `parallel` to make it simpler. Do you think it is OK? Otherwise it would have been `parallelism_min_length`.

Reply:
Yup that's perfect IMO, thanks !